sdk/tests/core/test_az_hfsm_pipeline.c

```diff
@@ -0,0 +1,376 @@
+static int pinbound0 = 0;
+static int sinbound0 = 0;
+static int sinbound1 = 0;
+static int sinbound2 = 0;
+static int sinbound3 = 0;
+static int soutbound0 = 0;
+static int soutbound1 = 0;
+
+static az_result P01Root(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = AZ_OK;
+  (void)me;
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      refp01root++;
+      break;
+
+    case AZ_HFSM_EVENT_EXIT:
+      refp01root--;
+      break;
+
+    case AZ_HFSM_EVENT_TIMEOUT:
+      timeout0++;
+      if (timeout0 > 1)
+      {
+        // To test timeout failure.
+        ret = AZ_ERROR_ARG;
+      }
+      break;
+
+    #ifdef TRANSPORT_MQTT_SYNC
+    case AZ_HFSM_PIPELINE_EVENT_PROCESS_LOOP:
+      loopout0++;
+      break;
+    #endif
+
+    case P_OUTBOUND_0:
+      poutbound0++;
+      break;
+
+    case S_INBOUND_0:
+      sinbound0++;
+      ret = az_hfsm_pipeline_send_inbound_event((az_hfsm_policy*)me, sinbound1_evt);
+      break;
+
+    case S_INBOUND_2:
+      sinbound2++;
+      ret = az_hfsm_pipeline_send_inbound_event((az_hfsm_policy*)me, sinbound3_evt);
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result P02Root(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = AZ_OK;
+  (void)me;
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      refp02root++;
+      break;
+
+    case AZ_HFSM_EVENT_EXIT:
+      refp02root--;
+      break;
+
+    case AZ_HFSM_EVENT_ERROR:
+      sinbound3++;
+      break;
+
+    case S_INBOUND_1:
+      sinbound1++;
+      break;
+
+    case S_INBOUND_3:
+      sinbound3++;
+      ret = AZ_ERROR_ARG;
+      break;
+
+    case S_OUTBOUND_1:
+      soutbound1++;
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_result P03Root(az_hfsm* me, az_hfsm_event event)
+{
+  int32_t ret = AZ_OK;
+  (void)me;
+
+  switch (event.type)
+  {
+    case AZ_HFSM_EVENT_ENTRY:
+      refp03root++; 
+      break;
+
+    case AZ_HFSM_EVENT_EXIT:
+      refp03root--;
+      break;
+
+    #ifdef TRANSPORT_MQTT_SYNC
+    case AZ_HFSM_PIPELINE_EVENT_PROCESS_LOOP:
+      loopin0++;
+      break;
+    #endif
+
+    case AZ_HFSM_EVENT_ERROR:
+      _az_PRECONDITION_NOT_NULL(event.data);
+      az_hfsm_event_data_error* test_error = (az_hfsm_event_data_error*) event.data;
+      if (test_error->error_type == AZ_ERROR_ARG)
+      {
+        timeout1++;
+      } else
+      {
+        ret = AZ_ERROR_NOT_IMPLEMENTED;
+      }
+      break;
+
+    case P_INBOUND_0:
+      pinbound0++;
+      break;
+
+    case S_OUTBOUND_0:
+      soutbound0++;
+      ret = az_hfsm_pipeline_send_outbound_event((az_hfsm_policy*)me, soutbound1_evt);
+      break;
+
+    default:
+      ret = AZ_HFSM_RETURN_HANDLE_BY_SUPERSTATE;
+      break;
+  }
+
+  return ret;
+}
+
+static az_hfsm_state_handler p01_get_parent_test(az_hfsm_state_handler child_state)
```

Comment:
Also, parameter name is incorrect: it's not the parent state, it's the parent handler - the state is usually composite. E.g. a composite state would be `root\started\connecting` with `root`, `started`, `connected` being handlers.